memtable flush error handling

bhalevy opened the pull request:
The series unifies memtable flush error handling into table::seal_active_memtable
following up on f6d9d6175fad11bc1dbf0aa06e587953c0ba8dad.

The goal here is to prevent an infinite retry loop as in #10498
by aborting on any error that is not bad_alloc.

Fixes #10498

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/623/

mikolajsieluzycki commented on utils/extremum_tracking.hh:
```diff
@@ -14,15 +14,16 @@
 namespace detail {
 
 template<typename T, typename Comparator>
+requires std::is_nothrow_copy_constructible_v<T> && std::is_nothrow_move_constructible_v<T>
```
Update uses copy assignment operator, we should also require std::is_nothrow_copy_assignable_v<T>. In most cases the tracker probably operates on simple objects, but probably we could change that assignment to

`_value = std::move(value);`

In which case we'd need a std::nothrow_move_assignable_v<T> as well.

bhalevy replied:
Good idea

bhalevy replied:
I looked into this in v2, but the value can't always be moved.
I reduced the number of copies though to the minimum.

mikolajsieluzycki commented on utils/logalloc.cc:
```diff
@@ -704,27 +704,27 @@ class segment_store {
     }
-    size_t new_idx_for_segment(segment* seg) {
+    size_t new_idx_for_segment(segment* seg) noexcept {
         auto i = find_empty();
```
Can we mark find_empty as noexcept as well since we're not using ExecutionPolicy and pointer comparison won't throw?

mikolajsieluzycki commented on utils/logalloc.cc:
```diff
@@ -704,27 +704,27 @@ class segment_store {
     }
-    size_t new_idx_for_segment(segment* seg) {
+    size_t new_idx_for_segment(segment* seg) noexcept {
         auto i = find_empty();
```
Can we mark find_empty as noexcept as well since we're not using ExecutionPolicy and pointer comparison won't throw?

bhalevy replied:
done in v2.

mikolajsieluzycki commented on utils/logalloc.cc:
```diff
@@ -1074,20 +1074,20 @@ void segment_pool::prime(size_t available_memory, size_t min_free_memory) {
 }
 
-void segment_pool::on_segment_compaction(size_t used_size) {
+inline void segment_pool::on_segment_compaction(size_t used_size) noexcept {
```
If we're inlining functions in a .cc file, can we just move the definition to the class? The language fluff is particularly visible on oneliners:

```
inline void on_memory_allocation(size_t size) noexcept;
inline void segment_pool::on_memory_allocation(size_t size) noexcept {
    _stats.memory_allocated += size;
}
```
vs 

`inline void on_memory_allocation(size_t size) noexcept { _stats.memory_allocated += size; }`

mikolajsieluzycki commented on utils/logalloc.cc:
```diff
@@ -2540,7 +2540,7 @@ region_group::execution_permitted() noexcept {
 
 future<>
-region_group::start_releaser(scheduling_group deferred_work_sg) {
+region_group::start_releaser(scheduling_group deferred_work_sg) noexcept {
```
It's probably outside of the scope of this change, but why does this function start with yield?

bhalevy replied:
e40fb438f5470234194848b6cb69439b4afe453f

bhalevy commented on utils/logalloc.cc:
```diff
@@ -1117,20 +1117,20 @@ static segment_pool& get_shard_segment_pool() noexcept {
 static thread_local segment_pool& shard_segment_pool = get_shard_segment_pool();
 
-void segment::record_alloc(segment::size_type size) {
+void segment::record_alloc(segment::size_type size) noexcept {
```
Some can, but not all of them due to keeping the header thin and independent as possible (using forward declaration rather than including full class definitions for  dependencies)

bhalevy commented:
In v2 (3d32a02)
- cleaned min_max tracker copies and concept requirements
- add noexcept for `find_empty`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/634/

bhalevy commented:
ping @mikolajsieluzycki

bhalevy commented on utils/logalloc.hh:
```diff
@@ -284,7 +284,11 @@ public:
     }
 
     void decrease_evictable_usage(region_heap::handle_type& r_handle) {
+      try {
```
Good catch.  It could be a remnant from an early draft.
I'll remove that.

Unfortunately boost heap implementation might throw in `decrease`

bhalevy replied:
Fixed in v3

bhalevy commented on replica/table.cc:
```diff
@@ -581,8 +581,15 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
+    std::exception_ptr e;
+    for (;;) {
+        try {
+            if (!op && !_memtables->back()->empty()) {
```
We want to call `add_memtable` exactly once.
If we fail in `_flush_barrier.start()` so `op` is still disengaged,
the memtable at the back (the one that was added by a successful `add_memtable`) is expected to be empty.

We can define a state enum class here to control which states we need to go through until they succeed.

bhalevy commented on replica/table.cc:
```diff
@@ -625,19 +625,18 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
+                // If the node is configured to not abort on internal error,
+                // but propagate it up the chain, we can't do anything reasonable
+                // at this point.
+                abort();
```
We have some unit tests that create malformed sstables but I don't think we have one to generate an invalid memtable that can't be flushed.  Would be a good idea to test that.

bhalevy commented:
In v3 (e9950eba5):
- cleaned up the const/noexcept parade
- removed inadvertent ignoring of exceptions in `decrease_evictable_usage` as per https://github.com/scylladb/scylla/pull/10691#discussion_r884625995
- Fixed `has_sstable_write_permit` implementation.
- Changed retry loop implementation in `table::seal_active_memtable` to a state machine
  - and moved the reacquisition of the write permit into the `try_flush` state before flushing the memtable so we can handle errors there too
- added abort on timeout
- improved `memtable_test.failed_flush_prevents_writes` with additional error injection sites to test all failures in the state machine
  - also fixed validation logic to rely on `min_memtable_timestamp()` rather than `active_memtable().empty()`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/755/

bhalevy commented:
In v4 (adb522e8b4)
- check for shutdown earlier in `seal_active_memtable` catch block to ignore errors like semaphore_broken that might happen when the table is stopped.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/764/

bhalevy commented:
Incidentally, the following CI failure shows that the current unit test is flaky since the assert for active_memtable().empty() is timing-depdendent:
https://jenkins.scylladb.com/job/releng/job/Scylla-CI/782/artifact/testlog/x86_64_debug/memtable_test.3.log
```
ERROR 2022-06-12 18:07:46,607 [shard 0] database - Flush failed due to low memory. Retrying again in 100ms
memtable_test: test/boost/memtable_test.cc:933: auto failed_flush_prevents_writes::run_test_case()::(anonymous class)::operator()(cql_test_env &) const: Assertion `eventually_true([&t]() { return t.active_memtable().empty(); })' failed.
Aborting on shard 0.
```

mikolajsieluzycki commented:
@tgrabiec @bhalevy  Is this the same thing as https://github.com/scylladb/scylla/commit/0e78ad50ea52f3aecb6380527da4152eeab0881e and should be fixed by moving the triggering of soft pressure inside the eventually_true loop?

bhalevy commented:
> @tgrabiec @bhalevy Is this the same thing as [0e78ad5](https://github.com/scylladb/scylla/commit/0e78ad50ea52f3aecb6380527da4152eeab0881e) and should be fixed by moving the triggering of soft pressure inside the eventually_true loop?

I don't understand.  It is in the eventually_true loop since 0e78ad50ea.
The problem I mentioned in https://github.com/scylladb/scylla/pull/10691#issuecomment-1153253714 is due to active_memtable().empty() being unreliable and timing dependent.

In this series I included https://github.com/scylladb/scylla/pull/10691/commits/c954629cf1133d7d8b4573d40414b2899665d32c
that check that the memtable was flushed more reliably.

mikolajsieluzycki commented:
I assume by "flush loop is already running" you mean the loop in dirty_memory_manager::flush_when_needed. It [waits](https://github.com/scylladb/scylla/blob/e87ca733f0d331611d33db9b540eb11d68c9ccff/replica/database.cc#L1618) until gets notified that there is pressure or shutdown. The moment we call notify_soft_pressure in the test, the signal _should_flush is notified, then we wait for flush permit. If at this point the LSA region gets updated, the pressure will be released, because there really isn't any (memory consumption is low, in the test we just triggered the signal). By the time we get to the [check](https://github.com/scylladb/scylla/blob/e87ca733f0d331611d33db9b540eb11d68c9ccff/replica/database.cc#L1626), there is no pressure and we don't trigger the flush and start sleeping again until pressure builds up. The test at this point doesn't call notify_soft_pressure anymore, so we wait for the empty memtable, which never happens.

Adding notify_soft_pressure in the loop means that even if LSA region gets updated and prevents the flush, we will still call notify_soft_pressure again and get another shot at flushing. I don't know if at this point there is a chance that the region will be updated, and if it can't, the change won't be necessary.

bhalevy commented:
> I assume by "flush loop is already running" you mean the loop in dirty_memory_manager::flush_when_needed. It [waits](https://github.com/scylladb/scylla/blob/e87ca733f0d331611d33db9b540eb11d68c9ccff/replica/database.cc#L1618) until gets notified that there is pressure or shutdown. The moment we call notify_soft_pressure in the test, the signal _should_flush is notified, then we wait for flush permit. If at this point the LSA region gets updated, the pressure will be released, because there really isn't any (memory consumption is low, in the test we just triggered the signal). By the time we get to the [check](https://github.com/scylladb/scylla/blob/e87ca733f0d331611d33db9b540eb11d68c9ccff/replica/database.cc#L1626), there is no pressure and we don't trigger the flush and start sleeping again until pressure builds up. The test at this point doesn't call notify_soft_pressure anymore, so we wait for the empty memtable, which never happens.
> 
> Adding notify_soft_pressure in the loop means that even if LSA region gets updated and prevents the flush, we will still call notify_soft_pressure again and get another shot at flushing. I don't know if at this point there is a chance that the region will be updated, and if it can't, the change won't be necessary.

Sorry, My mind is already set at this series, not the pre-existing state.
With it, there's only one loop (instead of two plus additional error handling in `try_flush_memtable_to_sstable`), in `seal_active_memtable`.

mikolajsieluzycki commented:
I think I get your point. As long as the error is injected, we'll loop inside seal_active_memtable. We won't leave an option to leave that function and enter flush_if_needed again. Once we stop injecting the error, the loop inside seal_active_memtable will go to completion and the table will be properly flushed. Since that loop doesn't depend on soft pressure at all, we don't even need to notify relief/pressure after we stop injecting the error. LGTM

bhalevy commented:
@scylladb/scylla-maint please expedite merging of the PR as CI fequently fails in the memtable error injection test, e.g. https://jenkins.scylladb.com/job/releng/job/Scylla-CI/804/artifact/testlog/x86_64_debug/memtable_test.3.log
and this PR fixes that.

nyh commented:
@bhalevy when a PR was heavily reviewed by someone already (in this case, @mikolajsieluzycki), it's good to ask him to "approve" the PR before mergine. Now I need to either read all his comments to try to understand if he was please at the end, or review the code myself - or both. I'll do it now.

nyh commented on replica/database.cc:
```diff
@@ -1647,31 +1647,34 @@ future<> dirty_memory_manager::flush_when_needed() {
                 // want to start a new one as soon as the permits are destroyed and the semaphore is
                 // made ready again, not when we are done with the current one.
-                (void)this->flush_one(mtlist, std::move(permit)).then([r]() {
+                (void)this->flush_one(mtlist, std::move(permit)).then_wrapped([this, r](future<> f) {
```
I don't understand why you wanted to change this code. The old code had then() and handle_exception() - the new code has one then_wrapped() for both. Did you do this because it's more efficient? Or what? It's not explained in the commit message.

nyh commented on dirty_memory_manager.hh:
```diff
@@ -28,12 +28,21 @@ class sstable_write_permit final {
 public:
-    sstable_write_permit(sstable_write_permit&&) noexcept = default;
-    sstable_write_permit& operator=(sstable_write_permit&&) noexcept = default;
+    sstable_write_permit(sstable_write_permit&& o) noexcept : _permit(std::exchange(o._permit, std::nullopt)) {}
```
The "=default" doesn't work any more? Or what's the motivation here?

bhalevy replied:
I'll add the motivation to the commit log.

The default move implementation calls std::optional move operator which moves the value contents but leaves the source optional engaged if it was already engaged (just containing a moved-from value).
This makes it harder to check for implementing the bool operator below, and be used for `has_sstable_write_permit`

avikivity replied:
In general, after a move one should not use the moved-from value. Maybe the caller should have used std::exchange() if it intended to call operator bool later?

bhalevy replied:
I think we can move the logic up a layer to `flush_permit` below.
so it can contain std::optional<sstable_write_permit> and can use std::exchange to release it and has_value() to check its presence.

nyh commented on dirty_memory_manager.hh:
```diff
@@ -28,12 +28,21 @@ class sstable_write_permit final {
         return sstable_write_permit();
     }
+
+    operator bool() const noexcept {
```
Not motivated in the commit message. I'm not thrilled about implicit conversions operators (except for specific kinds of uses), but I'm not strongly against either.

nyh commented on replica/table.cc:
```diff
@@ -642,24 +642,26 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
+                // If the node is configured to not abort on internal error,
+                // but propagate it up the chain, we can't do anything reasonable
+                // at this point.
+                abort();
```
Wow, this on_internal_error_noexcept is really confusing. Why does it even exist in this form? I'll open a separate issue about that. https://github.com/scylladb/seastar/issues/1095

nyh reviewed (approved):
Overall looks good, although with literally **hundreds** of noexcept changes, there's a high chance that if you had a bug in one of them I missed it :-( I still don't know why we need to use noexcept so much, but I guess this ship has already sailed. Let's hope that if we have noexcept crashes (a noexcept which isn't really noexcept and can throw, and now crash) we'll catch them in a test - but we can't really be sure.

I also left a few comments, but now, after reviewing 50 patches, I don't even remember what these comments were, so please take a look.

bhalevy commented on db_clock.hh:
```diff
@@ -33,13 +33,13 @@ public:
         return time_point(std::chrono::duration_cast<duration>(std::chrono::seconds(t)));
     }
-    static time_point now() {
+    static time_point now() noexcept {
```
The function above is `static constexpr` therefore I didn't consider noecept for it.

The decision of where to add noexcept in this series was based on what code paths are being called from memtable flush, to try to isolate the exact function calls that may throw and build the state machine in table::seal_active_memtable around them and the error injection points that simulate their failure.

bhalevy commented on utils/logalloc.cc:
```diff
@@ -748,9 +748,13 @@ class segment_store {
-        auto i = find_empty();
-        return i != _segments.end();
+    bool can_allocate_more_segments() const noexcept {
+        for (auto it = _segments.cbegin() + 1; it != _segments.cend(); ++it) {
```
`find_empty` is non-const.
Instead of open-coding the loop I can use:
```
        auto i = std::find(_segments.cbegin() + 1, _segments.cend(), nullptr);
        return i != _segments.cend();
```

not sure if it is any better, but it reflects a smaller change.

bhalevy replied:
> Can't this be just fixed?

We can add a const overload for find_empty, yes.

bhalevy commented on utils/logalloc.cc:
```diff
@@ -1088,20 +1088,20 @@ void segment_pool::prime(size_t available_memory, size_t min_free_memory) {
 }
 
-void segment_pool::on_segment_compaction(size_t used_size) {
+inline void segment_pool::on_segment_compaction(size_t used_size) noexcept {
```
These methods are marked inline in the class definition ass well, they are not free global functions so `inline` should apply, as if they were defined in the class definition body.

bhalevy replied:
Alternatively, we can also move the definition into the class definition, but this would create more churn. (can also be done as a follow up)

bhalevy commented on utils/logalloc.cc:
```diff
@@ -830,7 +830,10 @@ class segment_pool {
+    segment_descriptor& descriptor(segment* seg) noexcept {
+        uintptr_t index = idx_from_segment(seg);
+        return _segments[index];
+    }
```
it's just a step towards the next patch that adds a const overload of descriptor().
I thought it would help make reviewing of the patches a little easier.

bhalevy commented:
In v5 (eadbebe789)
- addressed @nyh review comments on v4
  - virtually no functional changes, only commentary in the commit logs and comments in the code
  - added const overload of find_empty, used in can_allocate_more_segments instead of open-coding the iteration.
  - made sstable_write_permit operator bool explicit.
  - changed use of `on_internal_error_noexcept` to simple logger.error followed by std::abort.
- rebased onto 5bd2e0ccce
- added "test: memtable_test: failed_flush_prevents_writes: notify_soft_pressure only once"

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/830/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1103/

bhalevy commented:
@nyh ping, please re-review.
I'm afraid the bits will rot quickly due to the breadth of the change.

bhalevy commented:
In v7 (46bc4ed318):
- rebased
- table: try_flush_memtable_to_sstable: propagate errors to seal_active_memtable
  - use coroutine::return_exception_ptr

bhalevy commented:
@nyh / @scylladb/scylla-maint ping

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1390/

bhalevy commented:
Looking into https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1390/artifact/testlog/x86_64/dev/database_test.test_truncate_without_snapshot_during_writes.114.log
```
WARN  2022-07-17 17:41:36,630 [shard 0] sstable - create_sharding_metadata: range=[{-468459073612751032, pk{00046b657930}}, {-468459073612751032, pk{00046b657930}}] has no intersection with shard=0 first_key={key: pk{00046b657930}, token:-468459073612751032} last_key={key: pk{00046b657930}, token:-468459073612751032} ranges_single_shard=[] ranges_all_shards={{1, {[{-468459073612751032, pk{00046b657930}}, {-468459073612751032, pk{00046b657930}}]}}}
ERROR 2022-07-17 17:41:36,630 [shard 0] table - failed to write sstable /jenkins/workspace/releng/Scylla-CI/scylla/testlog/x86_64/dev/scylla-e2b694c7-db4f-4f9d-9940-9c6c21850888/ks/cf-8f74aba005de11ed92fa8661a0ed7890/me-2-big-Data.db: std::runtime_error (Failed to generate sharding metadata for /jenkins/workspace/releng/Scylla-CI/scylla/testlog/x86_64/dev/scylla-e2b694c7-db4f-4f9d-9940-9c6c21850888/ks/cf-8f74aba005de11ed92fa8661a0ed7890/me-2-big-Data.db)
ERROR 2022-07-17 17:41:36,631 [shard 0] table - Memtable flush failed due to: std::runtime_error (Failed to generate sharding metadata for /jenkins/workspace/releng/Scylla-CI/scylla/testlog/x86_64/dev/scylla-e2b694c7-db4f-4f9d-9940-9c6c21850888/ks/cf-8f74aba005de11ed92fa8661a0ed7890/me-2-big-Data.db). Aborting, at 0x329e28e 0x329e780 0x329ea88 0xf5bc69 0xf956b1 0x3196dc4 0x3198037 0x319742a 0x32be2e4 0x32bd8e1 0x32ba01c 0x317f97d /lib64/libpthread.so.0+0x92a4 /lib64/libc.so.6+0x100322
```

I don't see how it can be related to the changes in this branch.  Trying to reproduce first.

bhalevy commented:
I couldn't reproduce this. Not with the same seed not with over 2000 runs with a random seed.

bhalevy commented:
> I couldn't reproduce this. Not with the same seed not with over 2000 runs with a random seed.

For the record, the test failure is fixed in https://github.com/scylladb/scylla/pull/11066

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1404/

avikivity commented on utils/logalloc.cc:
```diff
@@ -1987,11 +1987,15 @@ region& region::operator=(region&& other) {
+const region_group* region::group() const noexcept {
+    return get_impl().group();
+}
+
```
This is more than marking something noexcept.

bhalevy replied:
Right, sorry.  I'll separate it into a patch of its own.

bhalevy replied:
split out in v9

avikivity commented on replica/table.cc:
```diff
@@ -592,6 +592,8 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
     };
+    // Try flushing for up to 5 minutes
+    auto timeout = db::timeout_clock::now() + 300s;
     for (auto st = state::add_memtable; st != state::done;) {
```
5 minutes is a way too little, imagine a huge memtable filled with tombstones and/or a very slow disk.

abort on timeout should be extremely rare.

avikivity replied:
Maybe it's better to count the number of retries instead.

bhalevy replied:
Can do

bhalevy replied:
Done in v9

bhalevy commented on utils/logalloc.cc:
```diff
@@ -2034,17 +2034,17 @@ std::ostream& operator<<(std::ostream& out, const occupancy_stats& stats) {
-occupancy_stats tracker::impl::region_occupancy() noexcept {
-    reclaiming_lock _(*this);
+occupancy_stats tracker::impl::region_occupancy() const noexcept {
+    reclaiming_lock _(*const_cast<tracker::impl*>(this));
```
It does.  It touches `_reclaiming_enabled`.
Coming to think about it, `tracker::impl::region_occupancy` and `tracker::impl::occupancy` are synchronous.
@tgrabiec you added this in d20fae96a2da5d0fe381c91c42be6253503eb795
Why do we need the `reclaiming_lock` at all in those stats functions?

tgrabiec replied:
Reclaiming invalidates iterators into _regions, so it's a safety measure. In case of occupancy(), reclaiming should not happen because it doesn't currently allocate memory. I guess it was added there just in case.

bhalevy replied:
`reclaiming_lock` deleted in these functions in v9

bhalevy commented on replica/table.cc:
```diff
@@ -570,68 +571,58 @@ class database_sstable_write_monitor : public permit_monitor, public backlog_wri
-    old->region().ground_evictable_occupancy();
-    auto previous_flush = _flush_barrier.advance_and_await();
-    auto op = _flush_barrier.start();
+    // FIXME: over-indented in anticipation for error handling
```
np, I can keep the indentation as is in this patch and deal with it later.

bhalevy replied:
This patch is very hard to verify also due to moving other parts of the logic.
I'll try to see if I can implement it in a 2-3 steps so to make the transformation clearer to review.

bhalevy replied:
patch simplified in v9 with no indentation changes

bhalevy commented:
In v9 (05e0e1c5e9):
- logalloc: segment_store: mark functions noexcept: split out:
  - logalloc: segment_store, segment_pool: return const segment* from segment_from_idx() const
  - logalloc: segment_store, segment_pool: idx_from_segment: get a const segment* in const overload
- added logalloc: tracker::impl: occupancy: get rid of reclaiming_lock
  - as it's not needed and makes constifying {region_,}occupancy trivial
- logalloc: region_group: mark functions noexcept: split out:
  - logalloc: region_impl: add group() const overload
-  logalloc: region: mark functions const/noexcept: split out:
  - logalloc: region: add const group overload
- table: coroutinize seal_active_memtable
  - avoided indentation changes to make it more reviewable
  - also kept error handling logic (temporarily)
  - added table: reindent seal_active_memtable
- added dirty_memory_manager: flush_permit: make _sstable_write_permit optional
  - for simplifying dirty_memory_manager: flush_permit: add has_sstable_write_permit
- table: seal_active_memtable: abort on timeout: dropped in favor of
  - table: seal_active_memtable: abort if retried for too long

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1432/

avikivity commented on replica/table.cc:
```diff
@@ -600,38 +601,34 @@ table::seal_active_memtable(flush_permit&& permit) noexcept {
-  }
+        co_await std::move(previous_flush);
+    // keep `op` alive until after previous_flush resolves
+
```
This is about 12,875.3531 times simpler

avikivity commented on replica/table.cc:
```diff
@@ -580,49 +580,90 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
-    } catch (...) {
-        ex = std::current_exception();
+        tlogger.warn("Memtable flush failed due to: {}. Will retry in {}ms", e, r.sleep_time().count());
+        co_await r.retry();
```
Are you relying on this to provide the retry limit? It's a little implicit.

bhalevy replied:
Added FIXME comment about that in v10

avikivity commented on replica/table.cc:
```diff
@@ -580,49 +580,90 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
+                    // If the node is configured to not abort on internal error,
+                    // but propagate it up the chain, we can't do anything reasonable
+                    // at this point. The error is logged and we can try again later
+                }
```
This means we don't take the node down. Do we or don't we?

avikivity replied:
Ah, it's a preexisting problem (later solved).

bhalevy replied:
Added FIXME comment in v10

bhalevy commented on replica/table.cc:
```diff
@@ -654,6 +656,9 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
             } catch (const std::bad_alloc& e) {
+                if (allowed_retries-- <= 0) {
+                    abort_on_error();
+                }
```
I'll add FIXMEs in the next version to capture the intention.

bhalevy commented:
In v10 (731fa08e98):
- rebased
- dirty_memory_manager: flush_when_needed: move error handling to flush_one/seal_active_memtable
  - simplified by using a `with_retry` function instead of using a state machine
  - added temporary FIXME comments about further changes anticipated in the series

bhalevy commented on replica/table.cc:
```diff
@@ -580,49 +580,90 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
+                _config.cf_stats->pending_memtables_flushes_count++;
+                _config.cf_stats->pending_memtables_flushes_bytes += memtable_size;
+                [[fallthrough]];
+            case state::try_flush: {
```
Done in v10

bhalevy commented:
In v11 (bb9eddc67f):
- rebased on top of 81e20ceaab8f54938aa77842ee5c9d9e62639432 (that generated a lot of merge conflicts)
- retested
- dirty_memory_manager: region_group: make simple constructor noexcept
- dirty_memory_manager: region_group: mark functions noexcept
- dropped a couple of changes to function that the merge above deleted.
- table: try_flush_memtable_to_sstable: propagate errors to seal_active_memtable
  - propagate error upstream if _async_gate is closed and we have to stop retrying.
  - use deferred action for updating stats so they'll be updated also if `co_await with_retry` throws.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1537/
